sql/moz-fx-data-marketing-prod/ga_derived/downloads_with_attribution_v1/query.sql

```diff
@@ -0,0 +1,315 @@
+    clientId AS client_id,
+    visitId AS visit_id,
+    hit.appInfo.landingScreenName AS landing_page,
+    hit.page.pagePath AS pagepath,
+    CASE
+      WHEN (
+          hit.type = 'EVENT'
+          AND hit.eventInfo.eventAction = 'Firefox Download'
+          AND hit.eventInfo.eventCategory IS NOT NULL
+          AND hit.eventInfo.eventLabel LIKE 'Firefox for Desktop%'
+        )
+        THEN TRUE
+      ELSE FALSE
+    END AS has_ga_download_event,
+    hit.type AS hit_type
+  FROM
+    `moz-fx-data-marketing-prod.65789850.ga_sessions_*` AS ga
+  CROSS JOIN
+    UNNEST(hits) AS hit
+  WHERE
+    _TABLE_SUFFIX = FORMAT_DATE('%Y%m%d', @submission_date)
+),
+pageviews AS (
+  SELECT
+    client_id AS client_id,
+    visit_id AS visit_id,
+    COUNT(pagePath) AS pageviews,
+    COUNT(DISTINCT pagePath) AS unique_pageviews,
+  FROM
+    all_hits
+  WHERE
+    hit_type = 'PAGE'
+  GROUP BY
+    client_id,
+    visit_id
+),
+dl_events AS (
+  SELECT
+    client_id AS client_id,
+    visit_id AS visit_id,
+    ANY_VALUE(landing_page) AS landing_page,
+    LOGICAL_OR(has_ga_download_event) AS has_ga_download_event
+  FROM
+    all_hits
+  GROUP BY
+    client_id,
+    visit_id
+),
+ga_sessions AS (
+  SELECT
+    clientId AS client_id,
+    visitId AS visit_id,
+    ANY_VALUE(geoNetwork.country) AS country,
+    ANY_VALUE(trafficSource.adContent) AS ad_content,
+    ANY_VALUE(trafficSource.campaign) AS campaign,
+    ANY_VALUE(trafficSource.medium) AS medium,
+    ANY_VALUE(trafficSource.source) AS source,
+    ANY_VALUE(device.deviceCategory) AS device_category,
+    ANY_VALUE(device.operatingSystem) AS os,
+    ANY_VALUE(device.browser) AS browser,
+    ANY_VALUE(device.browserVersion) AS browser_version,
+    ANY_VALUE(device.language) AS `language`,
+    ANY_VALUE(totals.timeOnSite) AS time_on_site
+  FROM
+    `moz-fx-data-marketing-prod.65789850.ga_sessions_*` AS ga
+  WHERE
+    _TABLE_SUFFIX = FORMAT_DATE('%Y%m%d', @submission_date)
+  GROUP BY
+    client_id,
+    visit_id
+),
+ga_sessions_with_hits AS (
+  SELECT
+    *
+  FROM
+    ga_sessions ga
+  JOIN
+    pageviews p
+  USING
+    (client_id, visit_id)
+  JOIN
+    dl_events
+  USING
+    (client_id, visit_id)
+),
+stub AS (
+  SELECT
+    s.jsonPayload.fields.visit_id AS stub_visit_id,
+    s.jsonPayload.fields.dltoken AS dltoken,
+    (COUNT(*) - 1) AS count_dltoken_duplicates,
+    DATE(s.timestamp) AS download_date
+  FROM
+    `moz-fx-stubattribut-prod-32a5.stubattribution_prod.stdout` s
+  WHERE
+    DATE(s.timestamp) = @submission_date
+    AND s.jsonPayload.fields.log_type = 'download_started'
+  GROUP BY
+    stub_visit_id,
+    dltoken,
+    DATE(s.timestamp)
+),
+-- This will drop all the ga_sessions w/o a DLtoken but keep DLtoken without a GA session.
+-- This will also result in multiple rows is the ga.client_id is not unique for the day
+-- since this visit_id is missing from the stub.
+downloads_and_ga_session AS (
+  SELECT
+    gs.client_id AS client_id,
```

Comment:
We can `COALESCE(stub_visit_id, client_id) AS client_id` and remove the `stub_visit_id` from the SELECT and GROUP BY clauses

Reply:
I intentionally keep them as separate fields to identify cases where the `stub_visit_id` is not set vs no matching GA Session.